### ISSUE TITLE
Entropy Calculation

### ISSUE THREAD

**Issue by madhavajay** [DanyWind/fastai_bayesian_nn#1]
Hi,
I am a total beginner in this field, so I probably mis-understand how the entropy calculation works.

https://github.com/DanyWind/fastai_bayesian/blob/09bed15ea73921405a4d914c41bac2dea10c9bc9/fastai_bayesian/metrics.py#L8

I am wondering whats different and if there are any advantages to the method listed here:
http://www.cs.ox.ac.uk/people/yarin.gal/website/blog_3d801aa532c1ce.html

Is there any significance to calculating tau and including the weight decay term? Are these simply different "covariance functions"?
```
probs = []
for _ in xrange(T):
    probs += [model.output_probs(input_x)]
predictive_mean = numpy.mean(prob, axis=0)
predictive_variance = numpy.var(prob, axis=0)
tau = l**2 * (1 - model.p) / (2 * N * model.weight_decay)
predictive_variance += tau**-1
```

**Comment by dhuynh95** [comment]
Hello there ! :)

Well I just used the formula for entropy, which is - ∑ p log(p), except that in this case we do not only have one vector of probability with C classes, we have T samples of vectors of size C.
Therefore we have a vector of size T*C, and if we have n observations we have in total T*N*C values.

Because we are interested in the mean probability that our T samples give, we average them, before taking the entropy.

I don't know where you found this bit of code, but it seems like it's part of the KL term between the prior and the variationnal approximation, which seems to be a kind of something in equation (15) here : https://arxiv.org/pdf/1506.02142.pdf

I hope this will help ^^

**Comment by thomsonn** [comment]
Shouldn't the entropy be calculated over all samples, instead of just the means? In the paper, Gal estimates the uncertainty in the equation for the second raw moment (after equation 6) as a variance of forward passes. Using the mean ignores any spread in predictions over multiple draws.

**Comment by madhavajay** [comment]
How does it compare to this example:
https://gist.github.com/zlapp/e82902577aade2d296e52e0c26bd5ac0
```
def uncertainty(mcd_preds):
    mcd_probs = [probs for _,_,probs in mcd_preds]
    mcd_probs = torch.stack(mcd_probs)
    mcd_means = mcd_probs.mean(dim=0)
    mcd_std = mcd_probs.std(dim=0)
    _,idx = mcd_means.max(0)
    std = mcd_std.data[idx.item()].item()
    return std
```

**Comment by dhuynh95** [comment]
@thomsonn You could average over all the samples to have an average entropy for all your observations but you do not need to. 

Several estimators exist for the uncertainty of your model, Yarin Gal, Riashat Islam, and Zoubin Ghahramani have explored it here : http://www.cs.ox.ac.uk/people/yarin.gal/website/PDFs/NIPS_2016_workshop_paper.pdf

In the end, entropy and BALD are better estimates for the best observations to label than std of predictions which perform as well as random selection

**Comment by dhuynh95** [comment]
@madhavajay As said in my previous post, in the bit of code you showed me, he computes the uncertainty with the std which is not the best indicator, especially because the softmax will squash the probabilities and give irrelevant deviations.
